File: src/www/__test__/Post_20220715_HelloBlog_Context.js
import * as userSimulator from "./userSimulator";
import { mainView, queryAllByTestId, getByTestId } from "./queries";

export class Post_20220715_HelloBlog_Context {
  async goToTheBlogSection() {
    userSimulator.clickLink(mainView, "Blog");
    await userSimulator.waitForLoading();
  }

  async youShouldSeeAListOfPosts() {
    const posts = queryAllByTestId(mainView, "post-list--item");
    expect(posts).not.toHaveLength(0);
  }

  async theLastPostTitleShouldBeSThisPost(expected) {
    // expected = "Hello Blog"

    const posts = queryAllByTestId(mainView, "post-list--item");
    const last = posts.at(-1);

    expect(last).toHaveTextContent(expected);
  }

  async goToTheSPost(postTitle) {
    userSimulator.clickLink(mainView, postTitle);
    await userSimulator.waitForLoading();
  }

  async youShouldSeeTheSPost(the) {
    // the = "Hello Blog"

    const title = getByTestId(mainView, "post-title");
    expect(title).toHaveTextContent(the);
  }

  async thePostShouldContainSWhichIsHere(contain) {
    const body = getByTestId(mainView, "post-body");
    expect(body).toHaveTextContent(contain);
  }
}

File: src/test/java/com/drpicox/game/Post_20220715_HelloBlog_Context.java
package com.drpicox.game;

import com.drpicox.game.hints.api.ListPostsResponse;
import com.drpicox.game.hints.api.ListPostsResponseEntry;
import com.drpicox.game.hints.api.PostResponse;
import org.springframework.stereotype.Component;
import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth8.assertThat;
import com.drpicox.game.fixtures.FrontendSimulator;

import java.util.Optional;

@Component
public class Post_20220715_HelloBlog_Context {

    private final FrontendSimulator frontendSimulator;

    private ListPostsResponse postsList;
    private PostResponse post;

    Post_20220715_HelloBlog_Context(FrontendSimulator frontendSimulator) {
        this.frontendSimulator = frontendSimulator;
    }

    public void beforeTest() {
    }

    public void goToTheBlogSection() {
        postsList = frontendSimulator.get("/api/v1/posts", ListPostsResponse.class);
    }

    public void youShouldSeeAListOfPosts() {
        assertThat(postsList.getPosts()).isNotEmpty();
    }

    public void theLastPostTitleShouldBeSThisPost(String expected) {
        // expected = "Hello Blog"
        var query = findPostEntry(expected);
        assertThat(query).isPresent();
    }

    public void goToTheSPost(String the) {
        // the = "Hello Blog"
        var entry = findPostEntry(the);
        var id = entry.get().getId();
        post = frontendSimulator.get("/api/v1/posts/" + id, PostResponse.class);
    }

    public void youShouldSeeTheSPost(String the) {
        assertThat(post.getTitle()).isEqualTo(the);
    }

    public void thePostShouldContainSWhichIsHere(String contain) {
        assertThat(post.getBody()).contains(contain);
    }

    private Optional<ListPostsResponseEntry> findPostEntry(String expectedTitle) {
        return postsList.getPosts().stream().filter(p -> p.getTitle().equals(expectedTitle)).findAny();
    }

    public void afterTest() {
    }
}
